Teach TSan about Swift Task and Actor model

yln opened the pull request:
I have identified the following conceptual synchronization points at
which task data and computation can cross thread boundaries.  We need to
model these in TSan to avoid false positives:

Awaiting an async task (`AsyncTask::waitFuture`), which has two cases:
1) The task has completed (`AsyncTask::completeFuture`).  Everything
   that happened during task execution "happened before" before the
   point where we access its result.  We synchronize on the *awaited*
   task.
2) The task is still executing: the current execution is suspended and
   the waiting task is put into the list of "waiters".  Once the awaited
   task completes, the waiters will be scheduled.  In this case, we
   synchronize on the *waiting* task.

Note: there is a similar relationship for task groups which I still have
to investigate.  I will follow-up with an additional patch and tests.

Actor job execution (`swift::runJobInExecutorContext`):
Job scheduling (`swift::swift_task_enqueue`) precedes/happens before job
execution.  Also all job executions (switching actors or suspend/resume)
are serially ordered.

Note: the happens-before edge for schedule->execute isn't strictly
needed in most cases since scheduling calls through to libdispatch's
`dispatch_async_f`, which we already intercept and model in TSan.
However, I am trying to model Swift Task semantics to increase the
chance of things to continue to work in case the "task backend" is
switched out.

rdar://74256733

Commit:
Duplicate select Swift Concurrency tests for TSan

I am putting these tests into a new `Sanitizers/tsan` subfolder so we
can keep the same test file name (avoid need for `tsan-` prefix).  This
should increase the chance that changes to the tests make it to the TSan
duplicate.

I will move and rename existing sanitizer tests (including ASan) tests
to keep consistency in a follow-up PR.

Commit:
Enable TSan for copied tests

Commit:
actor_counters.swift: allow specifying of parameters

This is useful during debugging when manually running the test.

Commit:
Add support for calling TSan's acquire()/release()

Commit:
Teach TSan about Swift Task and Actor model

I have identified the following conceptual synchronization points at
which task data and computation can cross thread boundaries.  We need to
model these in TSan to avoid false positives:

Awaiting an async task (`AsyncTask::waitFuture`), which has two cases:
1) The task has completed (`AsyncTask::completeFuture`).  Everything
   that happened during task execution "happened before" before the
   point where we access its result.  We synchronize on the *awaited*
   task.
2) The task is still executing: the current execution is suspended and
   the waiting task is put into the list of "waiters".  Once the awaited
   task completes, the waiters will be scheduled.  In this case, we
   synchronize on the *waiting* task.

Note: there is a similar relationship for task groups which I still have
to investigate.  I will follow-up with an additional patch and tests.

Actor job execution (`swift::runJobInExecutorContext`):
Job scheduling (`swift::swift_task_enqueue`) precedes/happens before job
execution.  Also all job executions (switching actors or suspend/resume)
are serially ordered.

Note: the happens-before edge for schedule->execute isn't strictly
needed in most cases since scheduling calls through to libdispatch's
`dispatch_async_f`, which we already intercept and model in TSan.
However, I am trying to model Swift Task semantics to increase the
chance of things to continue to work in case the "task backend" is
switched out.

rdar://74256733

yln commented:
Note 1:
Here is another summary of the added happens-before edges for TSan:
```
AsyncTask::completeFuture(this) -> AsyncTask::waitFuture(this)  // wait on completed task
AsyncTask::completeFuture(waitingTask) -> AsyncTask::waitFuture(waitingTask)  // wait on unfinished task
swift::swift_task_enqueue(job) -> swift::runJobInExecutorContext(job)  // job scheduling
swift::runJobInExecutorContext(job) (end) -> swift::runJobInExecutorContext(job) (begin)  // job execution
```
@rjmccall @DougGregor @ktoso  does that make sense to you?  Can you audit the places where I am inserting those edges.


Note 2:
John also expressed the desire that we would like hooks that cannot only be used from TSan, but also other users to, e.g., customize the "task backend":
```
/// A hook to take over global enqueuing.
/// TODO: figure out a better abstraction plan than this.
SWIFT_EXPORT_FROM(swift_Concurrency) SWIFT_CC(swift)
void (*swift_task_enqueueGlobal_hook)(Job *job);
```

After discounting other options (weak definitions), I went back-and-forth on this quite a bit, but wasn't able to come up with something that felt general enough to be useful outside of the context of TSan.


Note 3:
I would like to land this PR quickly to get some tests running by CI and avoid regressions (since the Swift runtime is evolving so fast), but I promise to follow-up with:
* A solution and tests for task groups.
* Tests for "true positives" that demonstrate that TSan retains the ability to catch data races.  I can certainly come up with a few "stupid"/synthetic examples. @DougGregor: do we already have a sense what common pitfalls with Swift Concurrency could be?  Maybe this is less important than I initially thought because of the static compiler errors outlined in the new [proposal](https://github.com/apple/swift-evolution/blob/main/proposals/0302-concurrent-value-and-concurrent-closures.md)?

yln commented on stdlib/public/Concurrency/ThreadSanitizer.cpp:
```diff
@@ -0,0 +1,35 @@
+using TSanFunc = void(void *);
+
+void swift::_swift_tsan_acquire(void *addr) {
+  static auto ptr = (TSanFunc *)dlsym(RTLD_DEFAULT, "__tsan_acquire");
```
Is this okay to do here?  I noticed that the Swift runtime already uses `dlsym()` in other places.  Alternatively, I can turn this around and expose a `_swift_set_tsan_hooks()` and make TSan call that to avoid taking the `dlsym()` hit int the Swift runtime.

yln commented:
@swift-ci Please test

yln commented on stdlib/public/Concurrency/Task.cpp:
```diff
@@ -63,10 +63,12 @@ FutureFragment::Status AsyncTask::waitFuture(AsyncTask *waitingTask) {
       return queueHead.getStatus();
 
     case Status::Executing:
+      _swift_tsan_release(static_cast<Job *>(waitingTask));
```
One more implementation note: at this point, we actually can't guarantee yet that we are going to execute the `while (true)` loop with the `Executing` case.  The `if (... waitQueue.compare_exchange_weak ...)` below could fail and when we query the status again it could have changed.  So we are potentially adding a superfluous edge.  However, I don't think we want to put this call inside the `if` below.  This would mean that we put ourselves on the "wrong side of the race", i.e., when we loose this race we could produce a false positive (in very rare situations).

Commit:
Fix TSan symbol lookup on Windows

ktoso commented:
> Note: there is a similar relationship for task groups which I still have
to investigate. I will follow-up with an additional patch and tests.

Right yes, I can explain those a bit.

The moral equivalent of waiting on a future for a group is `swift_task_group_wait_next_throwing`, which suspends until at least one task submitted to the group completes (those complete with the usual completeFuture), and then unsuspends returning that value. It may immediately return with nil if no pending child tasks exist.

--- 

If I read those happens-before correctly (x -> (before) y) these look good.

--- 

> After discounting other options (weak definitions), I went back-and-forth on this quite a bit, but wasn't able to come up with something that felt general enough to be useful outside of the context of TSan.

I agree that would be quite useful... Even for being able to capture e.g. some metrics on time from complete to unsuspend waiter etc.

Would it not be expressed with things like 

```
SWIFT_EXPORT_FROM(swift_Concurrency) SWIFT_CC(swift)
void (*swift_task_before_completeTask_hook)(AsyncTask *task);
```

etc?

yln commented:
> Would it not be expressed with things like
> 
> ```
> SWIFT_EXPORT_FROM(swift_Concurrency) SWIFT_CC(swift)
> void (*swift_task_before_completeTask_hook)(AsyncTask *task);
> ```

We would need a hook at every point I have a `tsan_acquire/release`.  I can do that if we agree that this is the right way forward (and general enough to be useful).

kubamracek commented:
> * Tests for "true positives" that demonstrate that TSan retains the ability to catch data races.  I can certainly come up with a few "stupid"/synthetic examples. @DougGregor: do we already have a sense what common pitfalls with Swift Concurrency could be?  Maybe this is less important than I initially thought because of the static compiler errors outlined in the new [proposal](https://github.com/apple/swift-evolution/blob/main/proposals/0302-concurrent-value-and-concurrent-closures.md)?

Could you throw in just one "positive" test still in this PR, e.g. the simplest case of racing on a global variable? I think that would be a nice confirmation that the hooks aren't just silencing everything.

Commit:
Add positive TSan tests for Swift Concurrency

Add tests that check that TSan detects true positives in simple examples
for both actors and async tasks.

kubamracek commented:
> Add tests that check that TSan detects true positives in simple examples for both actors and async tasks.

Awesome, thank you!

Commit:
Use `swift_once` instead of C++ static initializers

yln commented:
@swift-ci Please test

yln commented on stdlib/public/Concurrency/ThreadSanitizer.cpp:
```diff
@@ -0,0 +1,57 @@
+#endif
+}
+
+swift::swift_once_t initOnceToken;
```
@mikeash We are now using `swift_once` instead of C++ static initializers.

yln commented:
> **Build failed**
> [Swift Test OS X Platform](https://ci.swift.org/job/swift-PR-macos/26130/)
> Git Sha - [99bcd87](https://github.com/apple/swift/commit/99bcd87af6effcde7429f1620cd44208741e3810)
Unrelated test:
```
Swift(watchsimulator-i386).IRGen/async.hop_to_executor.sil
```

yln commented:
> Swift Test Linux Platform

yln commented:
> **Build failed**
> [Swift Test Linux Platform](https://ci.swift.org/job/swift-PR-Linux/26105/)
> Git Sha - [99bcd87](https://github.com/apple/swift/commit/99bcd87af6effcde7429f1620cd44208741e3810)

Failing tests on Linux:
```
Swift(linux-x86_64).Sanitizers/tsan.actor_counters.swift  # will disable and investigate in follow-up
Swift(linux-x86_64).Sanitizers/tsan.racy_actor_counters.swift  # we need to manually demangle report on Linux
Swift(linux-x86_64).Sanitizers/tsan.racy_async_let_fibonacci.swift  # we need to manually demangle report on Linux
```

Commit:
Manually demangle TSan report to make test pass on Linux

Use `%swift-demangle` to make FileCheck against symbol names pass on
Linux.

Commit:
Mark test unsupported on Linux and Windows

A few other concurrency tests do so as well.  I will investigate once
those other tests have been resolved.

yln commented:
@swift-ci Please test

yln commented:
@mikeash: Is this good to go?

yln commented:
@swift-ci Please test

yln commented:
@swift-ci Please test

mikeash reviewed (approved):
Looks good. Were you also going to modify this to do the `dsym` dance inside the TSan so we don't pay for it in the common case? That would be best. Should be a minor mod on top of this.

Commit:
Add `swift_task_set_tsan_hooks()`

This change will be sequenced:
  1. Call this hook from the TSan runtime
  2. Remove `dlsym()` lookup from Swift runtime

yln commented on stdlib/public/Concurrency/ThreadSanitizer.cpp:
```diff
@@ -55,3 +55,9 @@ void swift::_swift_tsan_release(void *addr) {
   }
 }
+
+SWIFT_EXPORT_FROM(swift_Concurrency) SWIFT_CC(c)
+void swift_task_set_tsan_hooks(TSanFunc *acquire, TSanFunc *release) {
+  tsan_acquire = acquire;
+  tsan_release = release;
+}
```
Adopting this will be sequenced:
  1. Call this hook from the TSan runtime
  2. Remove `dlsym()` lookup from Swift runtime in a follow-up

```
➤ nm -g build/Ninja-RelWithDebInfoAssert+stdlib-DebugAssert/swift-macosx-x86_64//lib/swift/macosx/libswift_Concurrency.dylib | grep tsan
0000000000057a40 T _swift_task_set_tsan_hooks
```

Let me know if you have any remaining concerns.
@mikeash @kubamracek @devincoughlin

yln replied:
rdar://74256720 (Call swift_task_set_tsan_hooks() from TSan runtime)
rdar://75493372 (Remove dlsym() lookup for TSan functions)

mikeash replied:
Sounds fine to me, thanks! All LGTM.

yln commented:
@swift-ci Please test and merge

danliew-apple commented on stdlib/public/Concurrency/ThreadSanitizer.cpp:
```diff
@@ -0,0 +1,63 @@
+}
+
+swift::swift_once_t initOnceToken;
+void initializeThreadSanitizer(void *unused) {
```
I presume the `unused` here is to future proof the interface incase we need to pass data from TSan in the future?
